alchemist-engine/src/main/java/it/unibo/alchemist/core/EngineConfiguration.kt

```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright (C) 2010-2023, Danilo Pianini and contributors
+ * listed, for each module, in the respective subproject's build.gradle.kts file.
+ *
+ * This file is part of Alchemist, and is distributed under the terms of the
+ * GNU General Public License, with a linking exception,
+ * as described in the file LICENSE in the Alchemist distribution's top directory.
+ */
+
+package it.unibo.alchemist.core
+
+/**
+ * Contains the engine configuration parameters.
+ */
+interface EngineConfiguration
```

Comment:
Remind me why we need this instead of constructor params

Reply:
As per Arbitrary class loading system documentation:

`Alchemist is able to load arbitrary types conforming to the expected interface (or Scala trait). The expected type depends on where the class is requested.`

Configuration is loaded with
SimulationModel.visitBuilding<EngineConfiguration>(context, engineConfigurationDescriptor)

Which resolves to the specified type ex. EpsilonBatchEngineConfiguration

Reply:
But why cannot we instead have plain old constructor parameters and do:
`SimulationModel.visitBuilding<Simulation>(context, engineConfigurationDescriptor)`

It looks like `EngineConfiguration` is a container object that exists only to provide a single parameter to `Engine`

Reply:
I can see some issues with building the Simulation (and specifically its Engine implementation)

1. Simulation is a generics interface (Simulation<T, P extends Position<? extends P>>), how can we infer the generic types from yaml?

2. Engine has a non-simple type constructor signature:
`public Engine(final Environment<T, P> e, final long maxSteps, final Time t, final Scheduler<T> scheduler)`

Scheduler addition is new (with the intent of using composition to build a custom engine), so it can be ignored as we can construct a correct scheduler for the correct Engine type, but the Environment<T, P> is from the original implementation. Environment is not a simple properties container, it typically contains mutable objects instanced in some business logic.

> It looks like EngineConfiguration is a container object that exists only to provide a single parameter to Engine

There is a type hierarchy that progressively adds mandatory parameters:

`
interface BatchEngineConfiguration : EngineConfiguration {

    val outputReplayStrategy: String
}

class FixedBatchEngineConfiguration(
    override val outputReplayStrategy: String,
    val batchSize: Int,
) : BatchEngineConfiguration

class EpsilonBatchEngineConfiguration(
    override val outputReplayStrategy: String,
    val epsilonValue: Double,
) : BatchEngineConfiguration
`

So actually it supplies multiple parameters to the engine. Hence why I decided to map EngineConfiguration to an Engine impelentation instead of constructing an Engine from yaml.